src/Workspaces/Core/Portable/Workspace/Workspace.cs

```diff
@@ -795,6 +795,46 @@ internal void OnHasAllInformationChanged(ProjectId projectId, bool hasAllInforma
         }
 
         /// <summary>
+        /// Call this method when the document info changes, such as the name, folders or file path.
+        /// </summary>
+        protected internal void OnDocumentInfoChanged(DocumentId documentId, DocumentInfo newInfo)
+        {
+            using (_serializationLock.DisposableWait())
+            {
+                CheckDocumentIsInCurrentSolution(documentId);
+
+                var oldSolution = this.CurrentSolution;
+
+                var newSolution = oldSolution;
+                var oldInfo = oldSolution.GetDocument(documentId).GetDocumentInfoWithoutContent();
+
+                if (oldInfo.Name != newInfo.Name)
+                {
+                    newSolution = newSolution.WithDocumentName(documentId, newInfo.Name);
+                }
+
+                if (oldInfo.Folders != newInfo.Folders)
+                {
+                    newSolution = newSolution.WithDocumentFolders(documentId, newInfo.Folders);
+                }
+
+                if (oldInfo.FilePath != newInfo.FilePath)
+                {
+                    newSolution = newSolution.WithDocumentFilePath(documentId, newInfo.FilePath);
+                }
+
+                if (oldInfo.SourceCodeKind != newInfo.SourceCodeKind)
+                {
+                    newSolution = newSolution.WithDocumentSourceCodeKind(documentId, newInfo.SourceCodeKind);
+                }
+
+                SetCurrentSolution(newSolution);
+
+                this.RaiseWorkspaceChangedEventAsync(WorkspaceChangeKind.DocumentInfoChanged, oldSolution, newSolution, documentId: documentId);
```

Comment:
Hmm - general nitpick.  Should we avoid raising this if SetCurrentSolution ends up not actually doing anything?